File: Samples/FakeContext/FakeContext/TableFake.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using System.Linq.Expressions;
using System.Data.Linq;
using System.ComponentModel;

namespace FakeContext
{
    public class TableFake<TEntity> : ITable<TEntity>
        where TEntity : class
    {
        private readonly IList<TEntity> _entities;

        public TableFake(IList<TEntity> entities)
        {
            _entities = entities;
        }

        public IEnumerator<TEntity> GetEnumerator()
        {
            return _entities.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return _entities.GetEnumerator();
        }

        public Expression Expression
        {
            get { return _entities.AsQueryable().Expression; }
        }

        public Type ElementType
        {
            get { return _entities.AsQueryable().ElementType; }
        }

        public IQueryProvider Provider
        {
            get { return _entities.AsQueryable().Provider; }
        }

        public IQueryable CreateQuery(Expression expression)
        {
            return Provider.CreateQuery(expression);
        }

        public IQueryable<TElement> CreateQuery<TElement>(Expression expression)
        {
            return Provider.CreateQuery<TElement>(expression);
        }

        public object Execute(Expression expression)
        {
            return Provider.Execute(expression);
        }

        public TResult Execute<TResult>(Expression expression)
        {
            return Provider.Execute<TResult>(expression);
        }

        public void InsertOnSubmit(object entity)
        {
            Attach(entity);
        }

        public void InsertAllOnSubmit(IEnumerable entities)
        {
            AttachAll(entities);
        }

        public void Attach(object entity)
        {
            var item = entity as TEntity;
            if (item != null)
                _entities.Add(item);
        }

        public void Attach(object entity, bool asModified)
        {
            Attach(entity);
        }

        public void Attach(object entity, object original)
        {
            Attach(entity);
        }

        public void AttachAll(IEnumerable entities)
        {
            foreach (var entity in entities.OfType<TEntity>())
                _entities.Add(entity);
        }

        public void AttachAll(IEnumerable entities, bool asModified)
        {
            AttachAll(entities);
        }

        public void DeleteOnSubmit(object entity)
        {
            var item = entity as TEntity;
            if (item != null)
                _entities.Remove(item);
        }

        public void DeleteAllOnSubmit(IEnumerable entities)
        {
            foreach (var entity in entities.OfType<TEntity>())
                DeleteOnSubmit(entity);
        }

        public object GetOriginalEntityState(object entity)
        {
            return entity;
        }

        public ModifiedMemberInfo[] GetModifiedMembers(object entity)
        {
            return new ModifiedMemberInfo[0];
        }

        public DataContext Context
        {
            get { return null; }
        }

        public bool IsReadOnly
        {
            get { return false; }
        }

        public IList GetList()
        {
            return _entities.ToList();
        }

        public bool ContainsListCollection
        {
            get { return true; }
        }

        public void Attach(TEntity entity)
        {
            _entities.Add(entity);
        }

        public void Attach(TEntity entity, bool asModified)
        {
            Attach(entity);
        }

        public void Attach(TEntity entity, TEntity original)
        {
            Attach(entity);
        }

        public void AttachAll<TSubEntity>(IEnumerable<TSubEntity> entities) where TSubEntity : TEntity
        {
            foreach (var entity in entities.ToList())
                Attach(entity);
        }

        public void AttachAll<TSubEntity>(IEnumerable<TSubEntity> entities, bool asModified) where TSubEntity : TEntity
        {
            AttachAll(entities);
        }

        public void DeleteAllOnSubmit<TSubEntity>(IEnumerable<TSubEntity> entities) where TSubEntity : TEntity
        {
            foreach (var entity in entities.ToList())
                DeleteOnSubmit(entity);
        }

        public void DeleteOnSubmit(TEntity entity)
        {
            _entities.Remove(entity);
        }

        public ModifiedMemberInfo[] GetModifiedMembers(TEntity entity)
        {
            return new ModifiedMemberInfo[0];
        }

        public IBindingList GetNewBindingList()
        {
            return new BindingList<TEntity>(_entities);
        }

        public TEntity GetOriginalEntityState(TEntity entity)
        {
            return entity;
        }

        public void InsertAllOnSubmit<TSubEntity>(IEnumerable<TSubEntity> entities) where TSubEntity : TEntity
        {
            foreach (var entity in entities.ToList())
                Attach(entity);
        }

        public void InsertOnSubmit(TEntity entity)
        {
            Attach(entity);
        }
    }
}

File: Samples/FakeContext/FakeContext/TableFake.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Collections
Imports System.Linq.Expressions
Imports System.Data.Linq
Imports System.ComponentModel


Public Class TableFake(Of TEntity As Class)
    Implements ITable(Of TEntity)
    Private ReadOnly _entities As IList(Of TEntity)

    Public Sub New(ByVal entities As IList(Of TEntity))
        _entities = entities
    End Sub

    Public Function GetEnumerator() As IEnumerator(Of TEntity) Implements IEnumerable(Of TEntity).GetEnumerator
        Return _entities.GetEnumerator()
    End Function

    Private Function GetEnumerator2() As IEnumerator Implements IEnumerable.GetEnumerator
        Return _entities.GetEnumerator()
    End Function

    Public ReadOnly Property Expression() As Expression Implements ITable(Of TEntity).Expression
        Get
            Return _entities.AsQueryable().Expression
        End Get
    End Property

    Public ReadOnly Property ElementType() As Type Implements ITable(Of TEntity).ElementType
        Get
            Return _entities.AsQueryable().ElementType
        End Get
    End Property

    Public ReadOnly Property Provider() As IQueryProvider Implements ITable(Of TEntity).Provider
        Get
            Return _entities.AsQueryable().Provider
        End Get
    End Property

    Public Function CreateQuery(ByVal expression As Expression) As IQueryable Implements ITable(Of TEntity).CreateQuery
        Return Provider.CreateQuery(expression)
    End Function

    Public Function CreateQuery(Of TElement)(ByVal expression As Expression) As IQueryable(Of TElement) Implements ITable(Of TEntity).CreateQuery
        Return Provider.CreateQuery(Of TElement)(expression)
    End Function

    Public Function Execute(ByVal expression As Expression) As Object Implements ITable(Of TEntity).Execute
        Return Provider.Execute(expression)
    End Function

    Public Function Execute(Of TResult)(ByVal expression As Expression) As TResult Implements ITable(Of TEntity).Execute
        Return Provider.Execute(Of TResult)(expression)
    End Function

    Public Sub InsertOnSubmit(ByVal entity As Object) Implements ITable.InsertOnSubmit
        Attach(entity)
    End Sub

    Public Sub InsertAllOnSubmit(ByVal entities As IEnumerable) Implements ITable.InsertAllOnSubmit
        AttachAll(entities)
    End Sub

    Public Sub Attach(ByVal entity As Object) Implements ITable.Attach
        Dim item = TryCast(entity, TEntity)
        If item IsNot Nothing Then
            _entities.Add(item)
        End If
    End Sub

    Public Sub Attach(ByVal entity As Object, ByVal asModified As Boolean) Implements ITable.Attach
        Attach(entity)
    End Sub

    Public Sub Attach(ByVal entity As Object, ByVal original As Object) Implements ITable.Attach
        Attach(entity)
    End Sub

    Public Sub AttachAll(ByVal entities As IEnumerable) Implements ITable.AttachAll
        For Each entity In entities.OfType(Of TEntity)()
            _entities.Add(entity)
        Next
    End Sub

    Public Sub AttachAll(ByVal entities As IEnumerable, ByVal asModified As Boolean) Implements ITable.AttachAll
        AttachAll(entities)
    End Sub

    Public Sub DeleteOnSubmit(ByVal entity As Object) Implements ITable.DeleteOnSubmit
        Dim item = TryCast(entity, TEntity)
        If item IsNot Nothing Then
            _entities.Remove(item)
        End If
    End Sub

    Public Sub DeleteAllOnSubmit(ByVal entities As IEnumerable) Implements ITable.DeleteAllOnSubmit
        For Each entity In entities.OfType(Of TEntity)()
            DeleteOnSubmit(entity)
        Next
    End Sub

    Public Function GetOriginalEntityState(ByVal entity As Object) As Object Implements ITable.GetOriginalEntityState
        Return entity
    End Function

    Public Function GetModifiedMembers(ByVal entity As Object) As ModifiedMemberInfo() Implements ITable.GetModifiedMembers
        Return New ModifiedMemberInfo(-1) {}
    End Function

    Public ReadOnly Property Context() As DataContext Implements ITable.Context
        Get
            Return Nothing
        End Get
    End Property

    Public ReadOnly Property IsReadOnly() As Boolean Implements ITable.IsReadOnly
        Get
            Return False
        End Get
    End Property

    Public Function GetList() As IList Implements ITable(Of TEntity).GetList
        Return _entities.ToList()
    End Function

    Public ReadOnly Property ContainsListCollection() As Boolean Implements ITable(Of TEntity).ContainsListCollection
        Get
            Return True
        End Get
    End Property

    Public Sub Attach(ByVal entity As TEntity) Implements ITable(Of TEntity).Attach
        _entities.Add(entity)
    End Sub

    Public Sub Attach(ByVal entity As TEntity, ByVal asModified As Boolean) Implements ITable(Of TEntity).Attach
        Attach(entity)
    End Sub

    Public Sub Attach(ByVal entity As TEntity, ByVal original As TEntity) Implements ITable(Of TEntity).Attach
        Attach(entity)
    End Sub

    Public Sub AttachAll(Of TSubEntity As TEntity)(ByVal entities As IEnumerable(Of TSubEntity)) Implements ITable(Of TEntity).AttachAll
        For Each entity In entities.ToList()
            Attach(entity)
        Next
    End Sub

    Public Sub AttachAll(Of TSubEntity As TEntity)(ByVal entities As IEnumerable(Of TSubEntity), ByVal asModified As Boolean) Implements ITable(Of TEntity).AttachAll
        AttachAll(entities)
    End Sub

    Public Sub DeleteAllOnSubmit(Of TSubEntity As TEntity)(ByVal entities As IEnumerable(Of TSubEntity)) Implements ITable(Of TEntity).DeleteAllOnSubmit
        For Each entity In entities.ToList()
            DeleteOnSubmit(entity)
        Next
    End Sub

    Public Sub DeleteOnSubmit(ByVal entity As TEntity) Implements ITable(Of TEntity).DeleteOnSubmit
        _entities.Remove(entity)
    End Sub

    Public Function GetModifiedMembers(ByVal entity As TEntity) As ModifiedMemberInfo() Implements ITable(Of TEntity).GetModifiedMembers
        Return New ModifiedMemberInfo(-1) {}
    End Function

    Public Function GetNewBindingList() As IBindingList Implements ITable(Of TEntity).GetNewBindingList
        Return New BindingList(Of TEntity)(_entities)
    End Function

    Public Function GetOriginalEntityState(ByVal entity As TEntity) As TEntity Implements ITable(Of TEntity).GetOriginalEntityState
        Return entity
    End Function

    Public Sub InsertAllOnSubmit(Of TSubEntity As TEntity)(ByVal entities As IEnumerable(Of TSubEntity)) Implements ITable(Of TEntity).InsertAllOnSubmit
        For Each entity In entities.ToList()
            Attach(entity)
        Next
    End Sub

    Public Sub InsertOnSubmit(ByVal entity As TEntity) Implements ITable(Of TEntity).InsertOnSubmit
        Attach(entity)
    End Sub

    Public Overrides Function ToString() As String Implements ITable(Of TEntity).ToString
        Return MyBase.ToString()
    End Function
End Class

